rust: add `IoMem` abstraction.

wedsonaf opened the pull request:
This is based on what I described in LKML
(https://lore.kernel.org/lkml/YIbQ3dHOpyD%2FyymW@google.com/) back in
April.

This is in preparation for the PL061 driver to be merged.

Signed-off-by: Wedson Almeida Filho <wedsonaf@google.com>

ojeda commented on rust/kernel/io_mem.rs:
```diff
@@ -0,0 +1,195 @@
+            Self::check_offset::<$type_name>(offset);
+            let ptr = self.ptr.wrapping_add(offset);
+            // SAFETY: The type invariants guarantee that `ptr` is a valid pointer. The check above
+            // guarantees that the code won't link if `offset` makes the write go out of bounds
```
Nit: a more generic way to say it (which also matches the doc above and does not depend on how `build_error!` is implemented):

```suggestion
            // guarantees that the code won't build if `offset` makes the write go out of bounds
```

wedsonaf replied:
Fixed.

bjorn3 commented on rust/kernel/io_mem.rs:
```diff
@@ -0,0 +1,195 @@
+        }
+
+        // Try to map the resource.
+        // SAFETY: Just mapping the a memory range.
```
s/a//
maybe also mention that it is guaranteed to only be mapped once.

wedsonaf replied:
Fixed.

ojeda commented on rust/kernel/io_mem.rs:
```diff
@@ -0,0 +1,195 @@
+    pub(crate) fn new(
+        start: bindings::resource_size_t,
+        end: bindings::resource_size_t,
+    ) -> Option<Self> {
```
Should this be `Result`? i.e. if the callers pass the wrong arguments, it is supposed to be an error, or does the notion of a "no resource" make sense for some use cases?

wedsonaf replied:
It is possible for a device implementation(of a spec) not to have some optional register bank. I don't think we have enough information to decide what the appropriate error should be, so making this an `Option` we push the decision to the caller.

ojeda replied:
I was thinking more about the `checked_*` operations, not so much about `start == 0`.

wedsonaf replied:
Are you proposing that we return `Result<Option<Self>>` or `Result<Self>`?

ojeda replied:
If `start == 0` is expected but bad arguments are not, then `Result<Option<Self>>` might be the right fit semantically, yeah.

wedsonaf replied:
This is an internal API, used here: https://github.com/wedsonaf/linux/blob/pl061/rust/kernel/amba.rs#L174

I don't think making it `Result<Option<Self>>` is necessary because we don't surface this to developers, and I don't think we want to fail `probe` before even giving drivers a chance to run.

bjorn3 commented on rust/kernel/io_mem.rs:
```diff
@@ -0,0 +1,195 @@
+        /// Writes IO data to the given offset, known at compile time.
+        ///
+        /// If the offset is not known at compile time, the build will fail.
+        pub fn $name(&self, value: $type_name, offset: usize) {
```
Would it be possible to make `offset` a const generic? That way `Self::check_offset` can be evaluated at compile time and thus work even with optimizations disabled or with `#[inline]` on this function.

ojeda replied:
I assume someone will eventually ask for the non-const-generic-but-known-at-compile-time case (e.g. zeroing some register table using a loop). Or is there an easy way (i.e. without macros to unfold manually etc.) to e.g. use the index of a `for` loop within a static range to call a const generic?

wedsonaf replied:
I like the improvements we'd get, but I think the ergonomics wouldn't be good: to call it, we would need to do something like `mem.writeb::<X>(v)`, which is quite a mouthful and also changes the order of arguments when compared to C. This would be giving more ammunition to those unhappy with Rust with marginal extra benefit (when used correctly).

bjorn3 commented on rust/kernel/io_mem.rs:
```diff
@@ -0,0 +1,195 @@
+}
+
+impl Resource {
+    pub(crate) fn new(
```
This should be `unsafe`, right? `IoMem` depends on the fact that you can't duplicate a `Resource`.

wedsonaf replied:
I don't think `IoMem` depends on this for safety. Having the same range mapped multiple times should be ok; it may be bad for correctness, but why would it violate safety? We never use the returned pointer directly in Rust, only as arguments to readX/writeX.

I'm more concerned about the DMA case. My thinking was to make the DMA interface unsafe, but now that Gary mentioned it, it occurs to me that even a DMA "handle" is just a number that could be written to a device. IOW, what forces a driver to go through the unsafe DMA interface when it could just encode some arbitrary physical address and ask a bus mastering device to read/write it? Nothing at the moment, so this needs more thought.

nbdd0121 commented on rust/kernel/io_mem.rs:
```diff
@@ -0,0 +1,195 @@
+    /// The resource described by `res` is mapped into CPU's address space so that it can be
+    /// accessed directly. It is also consumed by this function so that it can't be mapped again
+    /// to a different address.
+    pub fn try_new(res: Resource) -> Result<Self> {
```
Shouldn't this be unsafe? The caller needs to guarantee that it "owns" the resource that `Resource` describes.

nbdd0121 replied:
Also, the IO memory mapped cannot trigger any DMA operation, otherwise DMA could be used to safely circumvent all Rust safety measures.

wedsonaf replied:
It's taking over the the resource. And the resource doesn't have a public constructor.

wedsonaf replied:
> Also, the IO memory mapped cannot trigger any DMA operation, otherwise DMA could be used to safely circumvent all Rust safety measures.

Of course we want drivers to be able to start DMA operations through this. What would the point of not allowing that?

The question is: how can we restrict this? My thinking was to leave DMA allocation safe, but getting a "handle" (that would be passed to a device for DMA) to the allocation would be unsafe with the restriction that the allocation should be kept alive until the device completed the DMA. (There's no way for us to know when a device is done, so this operation seems intrinsically unsafe.)

The problem with the strategy above is that a driver could write anything to the device, including arbitrary values as DMA handles, and that could of course violate memory safety. I suppose we could make this function unsafe with the requirement that callers need to be careful with DMA, but that's a big hammer and would make this function unnecessarily unsafe for devices like gpio that are not bus-mastering.

Any other ideas?

nbdd0121 replied:
I don't think it's too big an issue to make this function unsafe. A GPIO driver would only need to say: `// SAFETY: This device cannot perform DMA`.

wedsonaf replied:
Sure, but it's still a big hammer for bus-mastering devices though: "you need to be disciplined while programming operations involving DMA" is not very Rust-like, we'd be basically reverting to C mode and leaving it up to developers to ensure safety. I don't like that.

nbdd0121 replied:
It's up to the author of DMA-capable devices to ensure that all DMA operations are correct, and I don't think it could be checked. The best we can do is a separate `IoMem` that is safe to create but unsafe to access.

wedsonaf replied:
I suppose we could make the safety requirement be that either (a) no DMA operations are performed, or (b) DMA buffers should be retrieved via the `dma` module (which will have further requirements).

nbdd0121 replied:
Well, regardless how DMA buffers are retrieved, it's still up to the device to ensure that correct address is written isn't it?

wedsonaf replied:
It's up to the driver, yes.

But if we make the DMA "get handle" interface unsafe and require developers to go through it, we force them to consider the safety requirements then (when programming an operation). My feeling is that having the safety requirement only in `IoMem::try_new` that is buried in init code rarely looked at will allow developers to violate them without noticing it -- not too different from C.

nbdd0121 replied:
Maybe use macros to generate register maps would be the best approach for safety; some registers could be marked unsafe with this approach. E.g.
```rust
iomap! {
	unsafe PL061IoMem {
		  gpiodir: u8 @ 0x400,
		  gpiois: u8 @ 0x404,
		  gpiobe: u8 @ 0x408,
		  gpioiev: u8 @ 0x40C,
		  gpioie: u8 @ 0x410,
		  gpiomis: u8 @ 0x418,
		  gpioic: u8 @ 0x41C,
	}
}
```
and then for a DMA device a register could be written as
```rust
		unsafe dma_addr: u64 @ ...,
```

wedsonaf replied:
I like where this is going. It also goes well with Miguel's idea of automatically generating accessors for individual registers.

One problem with this approach is that we don't always encode the dma address via IO mem like this. For example, in the NVMe case, we use host memory to hold command queues -- the entries in these queues hold DMA addresses, and IO mem is only a doorbell to indicate that new entries have been added to queues.

nbdd0121 replied:
I have no experience with NVMe but from your description it is does not sound very different from how virtio queue works. The command queue setup registers could be made unsafe in that case, and it's then up to the command queue abstraction to propagate the safety requirements.

nbdd0121 commented on rust/kernel/io_mem.rs:
```diff
@@ -0,0 +1,195 @@
+    }
+
+    const fn check_offset<T>(offset: usize) {
+        if !Self::offset_ok::<T>(offset) {
+            crate::build_error!("IoMem offset overflow");
+        }
```
```suggestion
        crate::build_assert!(Self::offset_ok::<T>(offset), "IoMem offset overflow");
```

wedsonaf replied:
Better indeed, thanks!

nbdd0121 commented on rust/kernel/io_mem.rs:
```diff
@@ -0,0 +1,195 @@
+        }
+    }
+
+    define_read!(readb, try_readb, u8);
```
How about just name them `read_u8, read_u16, read_u32, read_u64`?

wedsonaf replied:
No, the names were chosen to match those of the kernel.

nbdd0121 replied:
Well, it's not always best to name them exactly as C does. `try_write_u32` certainly feels a lot more Rust-y than `try_writel`.

nbdd0121 replied:
I just made a mistake writing `writew` rather than `writel`; this is because RISC-V uses `b/h/w/d` while kernel uses `b/w/l/q` and I confused myself. One reason to prefer `u8/u16/u32/u64`!

wedsonaf replied:
I'm not claiming it's _always_ best. I'm claiming that this was an explicit and conscious choice for this interface. The idea is to reduce friction when writing drivers similar to existing C ones; in case you haven't seen it yet, here's the converted PL061 GPIO driver: https://lwn.net/Articles/863459/

wedsonaf replied:
> I just made a mistake writing `writew` rather than `writel`; this is because RISC-V uses `b/h/w/d` while kernel uses `b/w/l/q` and I confused myself. One reason to prefer `u8/u16/u32/u64`!

I'm sorry to hear that.

However, as discussed in the past: our target audience is kernel developers. At least for now, when we have to make a choice between being clearer for Rust or kernel developers, we'll choose kernel developers.

ojeda replied:
Gary, this is Linux code, not Rust code. Whether the names feel "Rust-y" or not is not really important -- what current kernel developers feel is. We need to be mindful of introducing naming changes in very well-known APIs, specially if the example driver is going to be compared with the C version side-by-side.

I prefer names like `_u32` myself too, but just because I think seeing the type is simpler. While it is tempting to improve naming, it should be discussed with others when we are in mainline, not before and on our own.

nbdd0121 replied:
Well, my confusion comes from the fact that I work with assembly and hardware, not that I am a Rust developer. I don't think it's really a friction for kernel developers to switch to `u8/u16/u32/u64`: they could easily understand that. Using `u8/u16/u32/u64` has an additional advantage that it's less friction for  new developers and people who work on lower levels that uses a different convention for sizes. I don't really see much advantage for keep using x86-centric `bwlq` convention.

nbdd0121 commented on rust/kernel/io_mem.rs:
```diff
@@ -0,0 +1,195 @@
+            // SAFETY: The type invariants guarantee that `ptr` is a valid pointer. The check above
+            // guarantees that the code won't link if `offset` makes the read go out of bounds
+            // (including the type size).
+            unsafe { concat_idents!(rust_helper_, $name)(ptr as _) }
```
Off-topic: Would be great if we have a `paste`-like macro

nbdd0121 reviewed (commented):
General API design question: do we want `readb`/`read_u8` or `read::<u8>()`?

wedsonaf commented:
### v1 -> v2

* Rebased
* Fixed build break on 32-bit archs (missing readq/writeq).
* Updated wording on comment (link vs build).
* Made `IoMem::try_new` unsafe.
* Fixed typo in safety comment when calling `rust_helper_ioremap`.
* Use `build_assert!` instead of `if !x { build_error!(...) }`.

wedsonaf commented:
@ojeda -- it looks like `ctypes::u_long` for arm (32 bits) is mapped to `u32` when building the kernel but `u64` in `rusttest`. The kernel builds fine, but `rusttest` is failing with the following (from https://github.com/Rust-for-Linux/linux/pull/462/checks?check_run_id=3196992364):

```
error[E0308]: mismatched types
   --> rust/kernel/io_mem.rs:161:61
    |
161 |         let addr = unsafe { rust_helper_ioremap(res.offset, res.size) };
    |                                                             ^^^^^^^^
    |                                                             |
    |                                                             expected `u64`, found `u32`
    |                                                             help: you can convert a `u32` to a `u64`: `res.size.into()`

error: aborting due to previous error
```

With `rust_helper_ioremap` defined as:

```
    fn rust_helper_ioremap(
        offset: bindings::resource_size_t,
        size: c_types::c_ulong,
    ) -> *mut c_types::c_void;
```

Is this expected? (Or am I misreading the error?)

fbq commented on rust/kernel/io_mem.rs:
```diff
@@ -0,0 +1,200 @@
+
+    const fn offset_ok<T>(offset: usize) -> bool {
+        if let Some(end) = offset.checked_add(core::mem::size_of::<T>()) {
+            if end <= SIZE {
```
I'm not 100% sure, but those `readX()` and `writeX()` seems to require the address being naturally aligned (i.e. aligned to the size of read and write), find something in the doc: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/driver-api/device-io.rst#n327

Won't hurt to get this clear.

wedsonaf replied:
Are you proposing that we check for natural alignment?

If it isn't a requirement for all platforms, I don't think we can do it here.

fbq replied:
> Are you proposing that we check for natural alignment?
> 

Yes, that's what's in my mind.

> If it isn't a requirement for all platforms, I don't think we can do it here.

I understand your point, but without the natural alignment checks, the Rust version `readX()` and `writeX()` seems `unsafe` to me.

fbq replied:
> > If it isn't a requirement for all platforms, I don't think we can do it here.
> 
> I understand your point, but without the natural alignment checks, the Rust version `readX()` and `writeX()` seems `unsafe` to me.

Except that `readb` and `writeb` are `safe`.

nbdd0121 replied:
> If it isn't a requirement for all platforms, I don't think we can do it here.

If it's a requirement for any platform, the driver shouldn't depend on it. Perhaps we should offer an `_unchecked` variant, and who uses it need to prove that they won't ever be ran on platforms without misalignment memory access support.

wedsonaf replied:
The document linked clearly states that "on many platforms" this is a requirements. Not all.

But I do like the idea of offering two versions. I'll make this one this the checked one, and we can add an unchecked one in the future if/when we find a case where we need this.

fbq replied:
Note that the link is the only evidence I found talking about the alignment requirement, and actually it was added just recently in this [patch](https://lore.kernel.org/lkml/20210402090542.131194-10-marcan@marcan.st/), I don't find many reviews on the doc. `readX()` and `writeX()` are heavily used APIs in kernel, I'd expect there are a few places (e.g. comments or docs) saying the requirement, but I only find one.

fbq replied:
FWIW, send an [email](https://lore.kernel.org/linux-arch/YQQr+twAYHk2jXs6@boqun-archlinux/T/#u) to linux-arch for some clarification.

ojeda replied:
> If it's a requirement for any platform, the driver shouldn't depend on it. Perhaps we should offer an `_unchecked` variant, and who uses it need to prove that they won't ever be ran on platforms without misalignment memory access support.

We could provide a safe one that only exists in the platforms where it is so -- that way people can freely use it for platform-dependent drivers, and if someone ever tries to copy-paste the code or port the module to another platform, they would get an error.

wedsonaf replied:
What we have now: a version with an unadorned name that checks alignment -- it is safe (in that it doesn't require an `unsafe` block and doesn't issue unaligned reads/writes) for anyone to use, but they'll get an error if they try some unaligned operation.

If we ever find a device whose driver requires unaligned access, we can add an unsafe version (whose name will be suffixed with `_unchecked`) that doesn't have the alignment check, and whose callers must be convinced that it is safe.

wedsonaf commented:
> @ojeda -- it looks like `ctypes::u_long` for arm (32 bits) is mapped to `u32` when building the kernel but `u64` in `rusttest`. The kernel builds fine, but `rusttest` is failing with the following (from https://github.com/Rust-for-Linux/linux/pull/462/checks?check_run_id=3196992364):

Opened #465 for this.

wedsonaf commented:
### v2 -> v3

* Added alignment check

wedsonaf commented:
### v3 -> v4

* Worked around #465
